source/extensions/filters/http/geoip/geoip_filter.h

```diff
@@ -0,0 +1,126 @@
+
+  Stats::Scope& scope_;
+  Runtime::Loader& runtime_;
+  Stats::StatNameSetPtr stat_name_set_;
+  const Stats::StatName stats_prefix_;
+  const Stats::StatName total_;
+  const Stats::StatName unknown_hit_;
+  bool use_xff_;
+  const uint32_t xff_num_trusted_hops_;
+  absl::optional<std::string> geo_country_header_;
+  absl::optional<std::string> geo_city_header_;
+  absl::optional<std::string> geo_region_header_;
+  absl::optional<std::string> geo_asn_header_;
+  absl::optional<std::string> geo_anon_header_;
+  absl::optional<std::string> geo_anon_vpn_header_;
+  absl::optional<std::string> geo_anon_hosting_header_;
+  absl::optional<std::string> geo_anon_tor_header_;
+  absl::optional<std::string> geo_anon_proxy_header_;
+  bool at_least_one_geo_header_configured_{false};
+};
+
+using GeoipFilterConfigSharedPtr = std::shared_ptr<GeoipFilterConfig>;
+
+class GeoipFilter : public Http::StreamDecoderFilter,
+                    public LookupCallbacks,
+                    public Logger::Loggable<Logger::Id::filter> {
+public:
+  GeoipFilter(GeoipFilterConfigSharedPtr config, DriverSharedPtr driver);
+  ~GeoipFilter() override;
+
+  // Http::StreamFilterBase
+  void onDestroy() override;
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap& trailers) override;
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override;
+  // LookupCallbacks
+  void onLookupComplete(const absl::optional<std::string>& lookup_result,
+                        const absl::optional<std::string>& geo_header) override;
+
+private:
+  void lookupGeolocationData(
+      const absl::optional<std::string>& geo_header,
+      const Network::Address::InstanceConstSharedPtr& remote_address,
+      void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
+                           const LookupCallbacks& callbacks,
+                           const absl::optional<std::string>& geo_header) const);
+  void lookupGeolocationAnonData(
+      const absl::optional<std::string>& geo_header,
+      const Network::Address::InstanceConstSharedPtr& remote_address,
+      void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
+                           const LookupCallbacks& callbacks,
+                           const absl::optional<std::string>& geo_header) const);
+  bool allLookupsComplete();
+  // Allow the unit test to have access to private members.
+  friend class GeoipFilterPeer;
+  GeoipFilterConfigSharedPtr config_;
+  Http::StreamDecoderFilterCallbacks* decoder_callbacks_{};
+  DriverSharedPtr driver_;
+  bool all_lookups_invoked_{false};
+  Envoy::Http::RequestHeaderMap* request_headers_ = nullptr;
```

Comment:
I think the Envoy way to do this now is
`OptRef<Http::RequestHeaderMap> request_headers_;`